TOPAZ4 interpolation

einola opened the pull request:
# TOPAZ4 interpolation

Fixes #880 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR (unfortunately) does two things:

 1. Address issue #880 by using scipy.interpolate to fill in areas that were land in the TOPAZ4 grid, but may be ocean in a more highly resolved mesh.
 2. Uses the new batch of TOPAZ4 files, which Tim W. has prepared for neXtSIM Lagrangian. I did this because the scipy interpolation routine was much less forgiving of errors that using the old files would introduce. I don't know why it works better with the new set of files, but since we're using them for neXtSIM Lagrangian now, we might as well switch here, too! I'm currently uploading the new files to the SUMMER storage so everyone in SASIP can access them.

---
# Test Description

The June '23 setup runs. I haven't updated the integration test, because Joe's reworking the python scripts needed for that into cdl files.

---
# Documentation Impact

None, but we should document better how to run the June '23 setup.

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Use scipy.interpolate.griddata instead of Tim's homebrew

This uses scipy.interpolate.griddata to interpolate from TOPAZ to an
arbitrary grid, ignoring the TOPAZ land mask and interpolating and
extrapolating where there's no TOPAZ data. It's still only for
make_initNH.py, but I'll make a separate python file for
topaz4_interpolate, so that both make_initNH.py and era5_topaz4_maker.py
can use it.

einola committed:
Move the interpolation functions and friends to interpolaters.py

It's only DRY in the case of topaz4_interpolate; the other functions
are only used in era5_topaz4_maker.py, but this way I've seperated file
handling and interpolation and rotation.

einola committed:
Forgot a print debug statement

It's gone now.

einola committed:
Change the land_mask option to boundary

It was always about open vs closed boundary anyway, but now it's more
explicit. The land mask is always read in from the 'mask' variable in
the --grid-file.

einola committed:
Switch to the new batch of TOPAZ4 files

There's a bit of history here. The short version is that I've had no end
of trouble with the old TOPAZ4 files. This culminated in this branch,
because the script would crash in an unpredictable fashion (at least
for me), showing fields that were full of nonsensical values. However,
Tim W. has produced a new batch of TOPAZ files, which is what we're
using for Lagrangian neXtSIM these days. So, I decided to try these
instead, and now everything works as expected.

joewallwork commented on run/interpolaters.py:
It's usually spelt 'interpolators'.

joewallwork commented on run/interpolaters.py:
```diff
@@ -0,0 +1,75 @@
+import pyproj
+
+# Returns TOPAZ data interpolated from the data grid and coordinates to the target grid and coordinates
+def topaz4_interpolate(target_lon, target_lat, data, data_x, data_y, proj_string):
```
Suggest making the preceding comment into a docstring. Similarly for the other functions
```suggestion
def topaz4_interpolate(target_lon, target_lat, data, data_x, data_y, proj_string):
    """
    Returns TOPAZ data interpolated from the data grid and coordinates to the target grid and coordinates
    """
```

joewallwork commented on run/era5_topaz4_maker.py:
```diff
@@ -56,115 +57,9 @@ def era5_source_file_name(field, unix_time, path):
 # Returns the file name that holds the TOPAZ data for a given field at a given time
-def topaz4_source_file_name(field, unix_time, path):
+def topaz4_source_file_name(unix_time, path):
     unix_tm = time.gmtime(unix_time)
```
It'd be useful to have a docstring that says what the two variables mean. For example, it's not totally clear what the path should be because another subdirectory gets joined in the function.

joewallwork replied:
Thanks for adding docstrings for the other functions - they're really useful. Please could you add something here, too, to make it clearer what the arguments mean?

joewallwork commented on run/interpolaters.py:
```diff
@@ -0,0 +1,75 @@
+
+    # The rotation of the basis vector can be computed as the azimuth of the
+    # great circle path from the location to the location of the new pole.
+    return np.arctan2(math.cos(pole_lat) * np.sin(delta_lon), np.cos(rlat) * math.sin(pole_lat) - np.sin(rlat) * math.cos(pole_lat) * np.cos(delta_lon))
```
We could do with applying auto-formatting to Python files, as well as C++. Probably best to do after #889 is merged, though. (Not for this PR.)

joewallwork commented on run/interpolaters.py:
```diff
@@ -0,0 +1,75 @@
+# 2023-03-28 Add a wrap-around for the ERA longitude. This is formally
+# incorrect when this function is used for TOPAZ data, but since the target
+# point would have to be out of bounds of the source, it is not so important.
+def bilinear(eyes, jays, data):
```
I haven't seen the `eyes` and `jays` notations before, I like them! However, I feel these argument names do need to be explained in the docstring, otherwise users could get quite confused.

joewallwork reviewed (changes requested):
I have a few requests, mainly related to adding docstrings for the functions that make it easier to understand what they are intended for and what arguments they accept. Also, I'd recommend using `interpolators` as the module name.

einola committed:
Correct spelling in the module name interpolators

einola committed:
Docstrings for interpolators.py functions

I used the reStructuredText format for the docstrings, because that's
what my IDE suggested. It seems to be widely used, so this should be
fine.

einola commented:
Thanks for the review @jwallwork23. Good points about the docstrings, and I've added them to all the functions in ``interpolators.py``. I wasn't too sure about which formatting to go for, but I guess this should be discussed in #891, rather than here.

Thanduriel commented:
I wanted to run the updated script but I could not find the new TOPAZ4 files. The link in the [readme](https://github.com/sasip-climate/catalog-shared-data-SASIP/blob/main/forcings/TOPAZ4-1d-Arctic.md) gives a 404. There is no TOPAZ directory in the [catalog](https://ige-meom-opendap.univ-grenoble-alpes.fr/thredds/catalog/meomopendap/extract/SASIP/model-forcings/ocean_forcing/catalog.html) besides the old one. Is the upload not completed yet?

einola commented:
There was a permissions problem with the data I uploaded. It's fixed now, and you should be able to download what you need @Thanduriel

joewallwork reviewed (changes requested):
> The June '23 setup runs. I haven't updated the integration test, because Joe's reworking the python scripts needed for that into cdl files.

Given that it's no longer clear that moving to CDL files is the best option (and is unlikely to be done soon in either case), could you update the integration test in this PR?

einola committed:
Docstrings for era5_topaz4_maker.py

Simple docstrings to describe the functions. See comment from Joe on
github.

einola committed:
Merge remote-tracking branch 'origin/develop' into issue880_topaz_interpolation

einola committed:
Merge remote-tracking branch 'origin/develop' into issue880_topaz_interpolation

# Conflicts:
#	run/make_initNH.py

einola commented:
> > The June '23 setup runs. I haven't updated the integration test, because Joe's reworking the python scripts needed for that into cdl files.
> 
> Given that it's no longer clear that moving to CDL files is the best option (and is unlikely to be done soon in either case), could you update the integration test in this PR?

Well, I can ... but I end up with 80k deletions and 80k additions. I don't know how @timspainNERSC made these files, but I wrote a shell script with ``sed``, ``ncdump``, and ``ncks`` doing the heavy lifting. In the end, I get slightly different values, resulting in a huge amount of deletions and additions. So, I don't think it's worth it. Strictly speaking, it doesn't need updating, anyway.

joewallwork reviewed (approved):
> Well, I can ... but I end up with 80k deletions and 80k additions. I don't know how @timspainNERSC made these files, but I wrote a shell script with `sed`, `ncdump`, and `ncks` doing the heavy lifting. In the end, I get slightly different values, resulting in a huge amount of deletions and additions. So, I don't think it's worth it. Strictly speaking, it doesn't need updating, anyway.

Fair enough! Indeed, it's probably not worth it then.

Thanks for the further docstring additions. I think this is good to go.
